File: Swift/Sources/Qibla.swift
//
//  Qibla.swift
//  Adhan
//
//  Copyright © 2018 Batoul Apps. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

import Foundation

public struct Qibla {
    /* The heading to the Qibla from True North */
    public let direction: Double

    public init(coordinates: Coordinates) {
        let makkah = Coordinates(latitude: 21.4225241, longitude: 39.8261818)

        /* Equation from "Spherical Trigonometry For the use of colleges and schools" page 50 */
        let term1 = sin(makkah.longitude.degreesToRadians() - coordinates.longitude.degreesToRadians())
        let term2 = cos(coordinates.latitude.degreesToRadians()) * tan(makkah.latitude.degreesToRadians())
        let term3 = sin(coordinates.latitude.degreesToRadians()) * cos(makkah.longitude.degreesToRadians() - coordinates.longitude.degreesToRadians())

        let angle = atan2(term1, term2 - term3)
        direction = angle.radiansToDegrees().unwindAngle()
    }
}

File: src/Qibla.js
import Coordinates from './Coordinates'
import { degreesToRadians, radiansToDegrees, unwindAngle } from './MathUtils'

export default function qibla(coordinates) {
    var makkah = new Coordinates(21.4225241, 39.8261818);

    // Equation from "Spherical Trigonometry For the use of colleges and schools" page 50
    var term1 = (
        Math.sin(degreesToRadians(makkah.longitude) -
        degreesToRadians(coordinates.longitude))
    );
    var term2 = (
        Math.cos(degreesToRadians(coordinates.latitude)) *
        Math.tan(degreesToRadians(makkah.latitude))
    );
    var term3 = (
        Math.sin(degreesToRadians(coordinates.latitude)) *
        Math.cos(degreesToRadians(makkah.longitude) -
        degreesToRadians(coordinates.longitude))
    );
    var angle = Math.atan2(term1, term2 - term3);

    return unwindAngle(radiansToDegrees(angle));
}
